File: backend/analytics.py
"""Server-side analytics: month-over-month trends from raw Socrata rows.

Ported from frontend/src/lib/analytics.ts. The results are attached to the
ContextObject so Claude can cite trend data in its synthesis.
"""

from __future__ import annotations

import re
from collections import defaultdict
from datetime import datetime, timezone
from typing import Any, Callable

from backend.models import AnalyticsSummary, TrendItem

_PERMIT_NORMALIZE = [
    ("EXPRESS", "EXPRESS PERMIT"),
    ("RENOVATION", "RENOVATION/ALTERATION"),
    ("ALTERATION", "RENOVATION/ALTERATION"),
    ("SIGN", "SIGNS"),
    ("NEW CONSTRUCTION", "NEW CONSTRUCTION"),
    ("WRECK", "WRECKING/DEMOLITION"),
    ("DEMOLITION", "WRECKING/DEMOLITION"),
    ("ELEVATOR", "ELEVATOR EQUIPMENT"),
    ("REINSTATE", "REINSTATE REVOKED PMT"),
    ("EASY PERMIT", "EASY PERMIT PROCESS"),
]


def _normalize_permit_type(raw: str) -> str:
    upper = (raw or "").upper()
    for keyword, label in _PERMIT_NORMALIZE:
        if keyword in upper:
            return label
    return re.sub(r"^PERMIT\s*[-–—]\s*", "", upper).strip() or "OTHER"


def _to_year_month(date_str: str) -> str | None:
    """Parse an ISO-ish date string and return 'YYYY-MM'."""
    if not date_str:
        return None
    try:
        dt = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
        return f"{dt.year}-{dt.month:02d}"
    except (ValueError, TypeError):
        return None


def _current_year_month() -> str:
    now = datetime.now(timezone.utc)
    return f"{now.year}-{now.month:02d}"


def _format_month(ym: str) -> str:
    """'2026-04' -> \"Apr '26\""""
    try:
        y, m = ym.split("-")
        dt = datetime(int(y), int(m), 1)
        return dt.strftime("%b '%y")
    except (ValueError, IndexError):
        return ym


def compute_trends(
    records: list[dict[str, Any]],
    get_date: Callable[[dict], str],
    get_category: Callable[[dict], str],
    max_categories: int = 8,
    capped: bool = False,
) -> tuple[list[TrendItem], str | None]:
    """Compute month-over-month trends. Returns (trends, period_label).

    ``capped``: the source query hit its row limit. Rows arrive date-DESC, so
    the cap truncates the OLDEST month mid-month — comparing against it
    fabricates an increase. Dropping that month keeps every remaining month
    complete (in a busy area 2,500 crime rows can span under two months, so
    this is not a theoretical case).
    """
    by_month: dict[str, dict[str, int]] = defaultdict(lambda: defaultdict(int))

    for r in records:
        ym = _to_year_month(get_date(r))
        if not ym:
            continue
        cat = get_category(r)
        by_month[ym][cat] += 1

    months = sorted(by_month.keys())
    if capped and months:
        del by_month[months[0]]
        months = months[1:]
    if len(months) < 2:
        return [], None

    current_cal = _current_year_month()
    if months[-1] == current_cal and len(months) >= 3:
        current_month = months[-2]
        prior_month = months[-3]
    else:
        current_month = months[-1]
        prior_month = months[-2]

    current_data = by_month[current_month]
    prior_data = by_month[prior_month]
    all_cats = set(current_data.keys()) | set(prior_data.keys())

    items: list[TrendItem] = []
    for cat in all_cats:
        curr = current_data.get(cat, 0)
        prev = prior_data.get(cat, 0)
        # prior 0 → the percentage is undefined, not "+100%". None renders as
        # "new" downstream; the counts themselves carry the magnitude.
        pct = None if prev == 0 else round(((curr - prev) / prev) * 100)
        items.append(TrendItem(
            category=cat,
            current_count=curr,
            prior_count=prev,
            change_pct=pct,
        ))

    items.sort(key=lambda t: t.current_count, reverse=True)
    period = f"{_format_month(current_month)} vs {_format_month(prior_month)}"
    return items[:max_categories], period


def compute_analytics(
    crime_rows: list[dict[str, Any]] | None = None,
    three11_rows: list[dict[str, Any]] | None = None,
    permit_rows: list[dict[str, Any]] | None = None,
    capped: dict[str, bool] | None = None,
) -> AnalyticsSummary:
    """Compute an AnalyticsSummary from raw map data rows.

    ``capped`` mirrors MapDataResponse.capped ("crimes"/"requests_311"/
    "building_permits") so a row-limited fetch doesn't fabricate trends from
    its truncated oldest month.
    """
    capped = capped or {}
    crime_trends = None
    three11_trends = None
    permit_trends = None
    period = None

    if crime_rows:
        crime_trends, period = compute_trends(
            crime_rows,
            get_date=lambda r: r.get("date", ""),
            get_category=lambda r: r.get("primary_type", "UNKNOWN"),
            capped=capped.get("crimes", False),
        )

    if three11_rows:
        three11_trends, p = compute_trends(
            three11_rows,
            get_date=lambda r: r.get("created_date", ""),
            get_category=lambda r: r.get("sr_type", "UNKNOWN"),
            capped=capped.get("requests_311", False),
        )
        period = period or p

    if permit_rows:
        permit_trends, p = compute_trends(
            permit_rows,
            get_date=lambda r: r.get("issue_date", ""),
            get_category=lambda r: _normalize_permit_type(r.get("permit_type", "")),
            capped=capped.get("building_permits", False),
        )
        period = period or p

    return AnalyticsSummary(
        crime_trends=crime_trends or None,
        three11_trends=three11_trends or None,
        permit_trends=permit_trends or None,
        trend_period=period,
    )

File: frontend/src/lib/analytics.ts
import i18n from "./i18n";

export interface TrendRow {
  category: string;
  currentCount: number;
  priorCount: number;
  /** null = prior month was 0 — the percentage is undefined (renders "new"). */
  changePercent: number | null;
  color: string;
}

export interface PieSlice {
  label: string;
  value: number;
  color: string;
}

function toYearMonth(dateStr: string): string | null {
  const d = new Date(dateStr);
  if (isNaN(d.getTime())) return null;
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  return `${y}-${m}`;
}

function getCurrentYearMonth(): string {
  const d = new Date();
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}`;
}

export function computeTrends<T>(
  records: T[],
  getDate: (r: T) => string,
  getCategory: (r: T) => string,
  getColor: (category: string) => string,
  capped = false,
): TrendRow[] {
  const byMonthAndCategory = new Map<string, Map<string, number>>();

  for (const r of records) {
    const ym = toYearMonth(getDate(r));
    if (!ym) continue;
    const cat = getCategory(r);
    if (!byMonthAndCategory.has(ym)) byMonthAndCategory.set(ym, new Map());
    const catMap = byMonthAndCategory.get(ym)!;
    catMap.set(cat, (catMap.get(cat) ?? 0) + 1);
  }

  let months = [...byMonthAndCategory.keys()].sort();
  // Rows arrive date-DESC; a row-capped fetch truncates the OLDEST month
  // mid-month, so comparing against it fabricates an increase — drop it.
  if (capped && months.length) {
    byMonthAndCategory.delete(months[0]);
    months = months.slice(1);
  }
  if (months.length < 2) return [];

  const currentCalendar = getCurrentYearMonth();
  let currentMonth: string;
  let priorMonth: string;

  if (months[months.length - 1] === currentCalendar && months.length >= 3) {
    currentMonth = months[months.length - 2];
    priorMonth = months[months.length - 3];
  } else {
    currentMonth = months[months.length - 1];
    priorMonth = months[months.length - 2];
  }

  const currentData = byMonthAndCategory.get(currentMonth) ?? new Map<string, number>();
  const priorData = byMonthAndCategory.get(priorMonth) ?? new Map<string, number>();

  const allCategories = new Set([...currentData.keys(), ...priorData.keys()]);
  const rows: TrendRow[] = [];

  for (const cat of allCategories) {
    const curr = currentData.get(cat) ?? 0;
    const prev = priorData.get(cat) ?? 0;
    // prior 0 → percentage undefined, not "+100%"; null renders as "new".
    const change = prev === 0 ? null : ((curr - prev) / prev) * 100;
    rows.push({
      category: cat,
      currentCount: curr,
      priorCount: prev,
      changePercent: change === null ? null : Math.round(change),
      color: getColor(cat),
    });
  }

  rows.sort((a, b) => b.currentCount - a.currentCount);
  return rows;
}

export function computePieSlices<T>(
  records: T[],
  getCategory: (r: T) => string,
  getColor: (category: string) => string,
): PieSlice[] {
  const counts = new Map<string, number>();
  for (const r of records) {
    const cat = getCategory(r);
    counts.set(cat, (counts.get(cat) ?? 0) + 1);
  }

  const slices: PieSlice[] = [];
  for (const [label, value] of counts) {
    slices.push({ label, value, color: getColor(label) });
  }

  slices.sort((a, b) => b.value - a.value);
  return slices;
}

export function getTrendMonthLabels<T>(
  records: T[],
  getDate: (r: T) => string,
  capped = false,
): { current: string; prior: string } | null {
  const months = new Set<string>();
  for (const r of records) {
    const ym = toYearMonth(getDate(r));
    if (ym) months.add(ym);
  }
  let sorted = [...months].sort();
  if (capped && sorted.length) sorted = sorted.slice(1); // mirror computeTrends
  if (sorted.length < 2) return null;

  const currentCalendar = getCurrentYearMonth();
  let currentMonth: string;
  let priorMonth: string;

  if (sorted[sorted.length - 1] === currentCalendar && sorted.length >= 3) {
    currentMonth = sorted[sorted.length - 2];
    priorMonth = sorted[sorted.length - 3];
  } else {
    currentMonth = sorted[sorted.length - 1];
    priorMonth = sorted[sorted.length - 2];
  }

  const fmt = (ym: string) => {
    const [y, m] = ym.split("-");
    const date = new Date(Number(y), Number(m) - 1);
    const locale = i18n.language === "es" ? "es-US" : "en-US";
    return date.toLocaleDateString(locale, { month: "short", year: "2-digit" });
  };

  return { current: fmt(currentMonth), prior: fmt(priorMonth) };
}
